core/src/main/java/org/apache/calcite/tools/RelBuilder.java

```diff
@@ -1978,6 +1978,21 @@ public RelBuilder join(JoinRelType joinType, RexNode condition,
     return this;
   }
 
+  /** Creates a {@link org.apache.calcite.rel.core.Correlate}
+   * with correlateId and requiredColumns. */
+  public RelBuilder correlate(JoinRelType joinType,
+      CorrelationId correlationId, ImmutableBitSet requiredColumns) {
+    Frame right = stack.pop();
+    final Frame left = stack.pop();
+    final RelNode join = correlateFactory
+        .createCorrelate(left.rel, right.rel, correlationId, requiredColumns, joinType);
+    final ImmutableList.Builder<Field> fields = ImmutableList.builder();
+    fields.addAll(left.fields);
+    fields.addAll(right.fields);
+    stack.push(new Frame(join, fields.build()));
+    return this;
+  }
+
```

Comment:
To be consistent with other methods in `RelBuilder` I don't think `requiredColumns` would be an `ImmutableBitSet`. It would be something a bit friendlier. Write some tests in RelBuilderTest and see which API is most user-friendly.

Reply:
I changed it to a list of integers(using varargs could be another option) and added a test case. I think this is a bit more user-friendly.

Reply:
The test you added is ```
RelNode root = builder.scan("EMP")
        .variable(v)
        .scan("DEPT")
        .filter(
            builder.equals(builder.field(0),
                builder.field(v.get(), "DEPTNO")))
        .correlate(JoinRelType.LEFT, v.get().id,
            Lists.newArrayList(
                SqlTypeUtil.findField(builder.peek(1).getRowType(), "DEPTNO")))
        .build();
```

I think that the test should be
```
RelNode root = builder.scan("EMP")
        .variable(v)
        .scan("DEPT")
        .filter(
            builder.equals(builder.field(0),
                builder.field(v.get(), "DEPTNO")))
        .correlate(JoinRelType.LEFT, v.get().id, builder.field(2, 0, "DEPTNO"))
        .build();
```

That's to say, the signature should be
```
public RelBuilder correlate(JoinRelType joinType,
      CorrelationId correlationId, RexNode... requiredFields)
```
and also, the overloaded method
```
public RelBuilder correlate(JoinRelType joinType,
      CorrelationId correlationId, Iterable<? extends RexNode> requiredFields)
```

Normally, as in this case, `requiredFields` is one `RexInputRef`. But it might be several expressions, and if so, `correlate` should create an extra `Project`. (See how `aggregate(GroupKey, Iterable<AggCall>)` uses a `Registrar` to project extra expressions.)

Reply:
Thanks for the suggestion. I updated signature. So, in correlate case, I think we should create an extra project on right relation if given fields are not simply RexInputRefs. Added a new test case for that as well.